### PR TITLE
kubeadm: include a phase runner for `join`

### PR DESCRIPTION
**Pull request by neolit123**
**What type of PR is this?**
/kind feature

**What this PR does / why we need it**:
- add a phase runner
- expose more joinData fields and getter methods
- add TODOs

this change is needed before we start working on the join phases.

**Which issue(s) this PR fixes**:

Refs https://github.com/kubernetes/kubeadm/issues/1204

**Special notes for your reviewer**:

**Does this PR introduce a user-facing change?**:

```release-note
NONE
```

/assign @fabriziopandini 
@kubernetes/sig-cluster-lifecycle-pr-reviews 
/priority important-soon

### CODE REVIEW

File: cmd/kubeadm/app/cmd/join.go
Comment (fabriziopandini) [inline cmd/kubeadm/app/cmd/join.go]: How does behave the runner together with args? I'm mostly concerned about [this]( https://github.com/kubernetes/kubernetes/blob/4531ce692a3557f2c1985409480095d7c884c341/cmd/kubeadm/app/cmd/phases/workflow/runner.go#L308)

File: cmd/kubeadm/app/cmd/join.go
Comment (fabriziopandini) [inline cmd/kubeadm/app/cmd/join.go]: @neolit123 
I think that the refactoring we are starting for introducing phases is a good opportunity to refactor this function, because `fetchInitConfigurationFromJoinConfiguration` that returns init cfg and tlsBootstrapCfg is kind of weird IMO

I think that we should have two separated methods

- `func (j *joinData) TlsBootstrapCfg() (*clientcmdapi.Config, error)` doing the discovery and sets the private `j.tlsBootstrapCfg` field
- `func (j *joinData) InitCfg() (*kubeadmapi.InitConfiguration, *clientcmdapi.Config, error)`, that uses j.TlsBootstrapCfg() 

Pushing things further, we can also change `fetchInitConfiguration` in fetching only the ClusterConfiguration (and not the ComponentConfigs, so we start addressing the "without kube-proxy" scenario as well

I let you choose what part of this should be part of this PR and what should be instead added to the tracking issue as a separated task

File: cmd/kubeadm/app/cmd/join.go
Comment (fabriziopandini) [inline cmd/kubeadm/app/cmd/join.go]: I think we comment this one (or remove) unless it is needed.

**Review by fabriziopandini — COMMENTED** [review]
@neolit123 thanks for this PR! 
I left one question + some comments, but IMO this is already in a good shape. Well done!

Adding in CC all the people that signed up for the rest of this effort so they are in the loop since the beginning
/cc @yago
/cc @ereslibre 
/cc @RA489

File: cmd/kubeadm/app/cmd/join.go
Comment (neolit123) [reply cmd/kubeadm/app/cmd/join.go]: did some basic tests and couldn't see a problem with the `cobra.NoArgs` of the sub command.
we should test again once there are actual phases...

File: cmd/kubeadm/app/cmd/join.go
Comment (neolit123) [reply cmd/kubeadm/app/cmd/join.go]: > Pushing things further, we can also change fetchInitConfiguration in fetching only the ClusterConfiguration (and not the ComponentConfigs), so we start addressing the "without kube-proxy" scenario as well

> I let you choose what part of this should be part of this PR and what should be instead added to the tracking issue as a separated task

i will leave this for a separate PR.

Comment (neolit123) [reply cmd/kubeadm/app/cmd/join.go]: > I think that we should aim to have two separated methods
done.

**Comment by neolit123** [comment]
PR updated.

File: cmd/kubeadm/app/cmd/phases/workflow/runner.go
Comment (neolit123) [inline cmd/kubeadm/app/cmd/phases/workflow/runner.go]: @fabriziopandini i had to do this to unblock the PR in terms of not having any phases yet.
otherwise with this check `BindToCommand()` returned early and the setup was incomplete, causing a panic in a `cmd.Flags()` call.

i think this removal is fine, runner wise.

File: cmd/kubeadm/app/cmd/join.go
Comment (MalloZup) [inline cmd/kubeadm/app/cmd/join.go]: `nit` should be here `init` -> `join`?

Comment (neolit123) [reply cmd/kubeadm/app/cmd/join.go]: +1

File: cmd/kubeadm/app/cmd/join.go
Comment (neolit123) [reply cmd/kubeadm/app/cmd/join.go]: we are overwriting the variable in `j` directly.

**Comment by neolit123** [comment]
`verify-generated-docs.sh` wants attention too :roll_eyes:

File: cmd/kubeadm/app/cmd/phases/workflow/runner.go
Comment (neolit123) [reply cmd/kubeadm/app/cmd/phases/workflow/runner.go]: outdated.

File: cmd/kubeadm/app/cmd/join.go
Comment (yagonobre) [reply cmd/kubeadm/app/cmd/join.go]: Sorry @neolit123. I don't get this, we need to explicitly declare the err var?

**Review by yagonobre — APPROVED** [review]
Thanks @neolit123 
/lgtm
/hold
feel free to remove the hold

File: cmd/kubeadm/app/cmd/join.go
Comment (MalloZup) [reply cmd/kubeadm/app/cmd/join.go]: yop this was i was meaning. sorry for beeing unclear @neolit123 .

Comment (neolit123) [reply cmd/kubeadm/app/cmd/join.go]: updated to reduce the confusion.

Comment (MalloZup) [reply cmd/kubeadm/app/cmd/join.go]: Thx

**Review by MalloZup — APPROVED** [review]

File: cmd/kubeadm/app/cmd/phases/workflow/runner.go
Comment (fabriziopandini) [inline cmd/kubeadm/app/cmd/phases/workflow/runner.go]: If we are moving this here, please remove it from L#396

**Review by fabriziopandini — APPROVED** [review]
@neolit123 only one last small nit and then ready to lgtm!
/approve

File: cmd/kubeadm/app/cmd/phases/workflow/runner.go
Comment (neolit123) [reply cmd/kubeadm/app/cmd/phases/workflow/runner.go]: looks like i copied instead of moving.

**Pull request merged**
